Add functions for darwin-framework-tool

seokhee-lee opened the pull request:
This change add "read event" function for darwin.
It is needed for Matter Certification of controller to pass the test case for "read event".

read-event-by-id command is added in darwin-framework-tool for test:
-./darwin-framework-tool doorlock read-event-by-id 1 1 1

readEventsWithEndpointID API is added in MTRBaseDevice.h of matter framework for app development:
-[device readAttributesWithEndpointID:[NSNumber numberWithUnsignedShort:endpointId]
                               clusterID:[NSNumber numberWithUnsignedInteger:mClusterId]
                             attributeID:[NSNumber numberWithUnsignedInteger:mAttributeId]
                                  params:params
                                   queue:callbackQueue
                              completion:^(...);}];

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=24057) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/project-chip/connectedhomeip?pullRequest=24057) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/project-chip/connectedhomeip?pullRequest=24057) it.</sub>

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.h:
```diff
@@ -302,6 +302,30 @@ extern NSString * const MTRArrayValueType;
+ * device will be read.
+ *
+ * A non-nil eventID along with a nil clusterID will only succeed if the
+ * event ID is for a global event that applies to all clusters.
```
There is no such thing as a "global event".

seokhee-lee replied:
I removed the comments related to "global event".

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.h:
```diff
@@ -302,6 +302,30 @@ extern NSString * const MTRArrayValueType;
+ * event ID is for a global event that applies to all clusters.
+ */
+
+- (void)readEventsWithEndpointID:(NSNumber * _Nullable)endpointID
```
Why do we want to add this API?  What are the use cases?

seokhee-lee replied:
We are preparing Matter Certification for a Matter controller iOS app which can use this API to pass test cases related to "read event".

bzbarsky-apple replied:
@seokhee-lee Does the app actually "read events" one at a time like this (as opposed to subscribing to events)?  Or is this just so it can pass the relevant certification tests?  I ask, because those tests shoud be conditioned on PICS that say whether the app actually does this....

bzbarsky-apple replied:
@seokhee-lee I'm just trying to understand what the use cases are for "read event", mostly.  But if you're planning to do it, we can certainly add API for it in the framework; we just need to get rid of the code duplication to make this more maintainable.

seokhee-lee replied:
@bzbarsky-apple There are several "read event" test cases which we are considering for certification. One of the test cases is DRLK-3.3 in 22-27353-001_Matter-1.0-Application-Cluster-TestPlans which can give some information to understand what the use cases are for "read event". We are also reviewing the codes to be more maintainable and will update them if there are some modifications to enhance it.

bzbarsky-apple replied:
Thank you for the pointer to that test plan.  That test plan .... does not use "read event" as far as I can see.  As best I can tell, it's assuming (but not stating) that the client has a subscription to the server, so when the server generates events they are reported to the client...

seokhee-lee replied:
In our experience (we got the controller certifications for webOS and Android), tester requested logs of read event results of server (mandatory) and client (optional) for the test case. Unlike read attribute, read event returns all history of the events but not only the value at that point in time.

bzbarsky-apple replied:
@cjandhyala could you please take a look?  That does _not_ seem like a reasonable requirement.

cjandhyala replied:
@seokhee-lee you don't need readEvent per say, The purpose of TC is the client able to receive the events, on the Darwin platform you can subscript to all events ('subscribe-all-events ')  and filter for the StartUp and other events you are interested to look for. On the example app (TH) you can trigger events using the commands mentioned at https://github.com/project-chip/connectedhomeip/blob/e56c2ce42f4bf5d7a5aab8898174b4f14e7888e6/examples/all-clusters-app/linux/README.md

For DUT as client TCs, The chip-tool commands provided in the verification steps doc are for example on how the same functionality verified using chip-tool. DUTs can run any equivalent command that verifies the purpose of the test.

seokhee-lee replied:
@bzbarsky-apple , @cjandhyala , thanks for your supports. I think that there is some misunderstanding in our side. I will discuss more with our app development team on whether they have the other requirements except for certification. If it was just for certification, I will close this pull request.

bzbarsky-apple replied:
Just to be clear: this is totally worth adding if people are going to use it.  We just want to share the code with the attribute read setup as much as possible.

seokhee-lee replied:
@bzbarsky-apple I checked the requirement on read/subscribe event API from our app team. They said that "subscribing all event" has some performance issues and so our iOS Matter Controller app will read/subscribe some selective events by using the proposed APIs. I am not sure whether this reason is enough or not in open source side. If the reason is not enough to be merged in upstream, We will manage this patch in our local build system.

bzbarsky-apple replied:
> so our iOS Matter Controller app will read/subscribe some selective events by using the proposed APIs

OK. That seems fine, as long as you realize the proposed API will in practice only allow you to subscribe to three event paths reliably (that's how many subscriptions a device is required to support per fabric).

Given that, I think the only outstanding issue is the code duplication compared to the "subscribe attribute" bit.

seokhee-lee replied:
@bzbarsky-apple "subscribe to one event" already has been implemented in our side and we will add "subscribe to one event" in this PR soon. Now, attribute in darwin matter framework also seems not to support "read/subscribe selected ones". I think that it can make app developing more convenient. We will try to do PR for it after this PR.

seokhee-lee replied:
@bzbarsky-apple "subscribe to one event" is added in this PR. Please review again and let me know if more modifications are needed.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1464,6 +1464,185 @@ + (id)CHIPEncodeAndDecodeNSObject:(id)object
     return decodedData.GetDecodedObject();
 }
 
+template <typename DecodableEventType> class BufferedReadEventCallback final : public app::ReadClient::Callback {
```
This is duplicating a lot of code that existing attribute-reading things have.  There was no way to share the code?

Also, why is this a template if it's only ever instantiated with one type for the template parameter?

seokhee-lee replied:
I agree your opinion that we should reduce code duplication. However in order to remove the duplication of code, lots of code changes are needed in the existing codes. So for now, I think that the separated codes can help to implement the read event function quickly and that code refactoring work can be done with more verification and discussion. The reason to use template is that I think to make the refactoring work easier.

bzbarsky-apple replied:
@seokhee-lee I appreciate your desire to add this quickly; are you able to commit to doing the needed refactoring afterward?

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.h:
```diff
@@ -157,6 +158,21 @@ typedef void (^MTRDeviceConnectionCallback)(MTRBaseDevice * _Nullable device, NS
+/**
+ * Discover commissionable nodes.
+ */
+- (void)discoverCommissionableNodes;
```
Why is this API desirable?  Especially since it does not actually do what the name says, in that it does only DNS-SD discovery.

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.h:
```diff
@@ -157,6 +158,21 @@ typedef void (^MTRDeviceConnectionCallback)(MTRBaseDevice * _Nullable device, NS
+ *
+ * @param[in] queue The queue on which the callbacks will be delivered
+ */
+- (void)setDeviceDiscoveryDelegate:(id<MTRDeviceDiscoveryDelegate>)delegate
```
Why do we have a separate delegate setter here, instead of having the "discover things" function take a block to call?

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.mm:
```diff
@@ -93,6 +95,9 @@ @interface MTRDeviceController ()
 @property (readonly) os_unfair_lock deviceMapLock; // protects nodeIDToDeviceMap
+@property (readonly) chip::Dnssd::ResolverProxy mDNSResolver;
+@property (readonly) MTRDeviceDiscoveryDelegateBridge * deviceDiscoveryDelegateBridge;
+@property (readonly) chip::Dnssd::ResolverDelegateProxy resolverDelegagteProxy;
```
Spelling?  Also, where is this used?

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.mm:
```diff
@@ -120,7 +125,13 @@ - (instancetype)initWithFactory:(MTRDeviceControllerFactory *)factory queue:(dis
         }
         _operationalCredentialsDelegate->setChipWorkQueue(_chipWorkQueue);
+
+        _deviceDiscoveryDelegateBridge = new MTRDeviceDiscoveryDelegateBridge();
```
If this is going to get checked for null, doesn't the `new` need to be `std::nothrow`?

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.mm:
```diff
@@ -807,6 +818,14 @@ - (void)asyncDispatchToMatterQueue:(void (^)(chip::Controller::DeviceCommissione
 
+- (void)discoverCommissionableNodes
+{
+    _mDNSResolver.Shutdown(); // reset if already inited
+    _mDNSResolver.Init(chip::DeviceLayer::UDPEndPointManager());
+    _mDNSResolver.SetCommissioningDelegate(_deviceDiscoveryDelegateBridge);
+    chip::Dnssd::DiscoveryFilter filter(chip::Dnssd::DiscoveryFilterType::kNone, (uint64_t) 0);
+    _mDNSResolver.DiscoverCommissionableNodes(filter);
```
Pretty sure this can lead to thread data races, since it's going to touch the DNS service bits on some random caller thread.

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceController.mm:
```diff
@@ -866,6 +885,13 @@ - (void)invalidateCASESessionForNode:(chip::NodeId)nodeID;
+
+- (void)setDeviceDiscoveryDelegate:(id<MTRDeviceDiscoveryDelegate>)delegate queue:(dispatch_queue_t)queue
+{
+    VerifyOrReturn([self checkIsRunning]);
```
Silent no-op is not a great API....

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRDeviceDiscoveryDelegate.h:
```diff
@@ -0,0 +1,39 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef struct DiscoveredNodeData {
```
Does this get reflected into Swift correctly?

seokhee-lee replied:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification.

bzbarsky-apple reviewed (changes requested):
I'd like to understand why we are adding these things to darwin-framework-tool and the framework.

In particular, the sequence here should be figuring out what we want in the framework and why, then adding it to the framework for those reasons, then adding a way for darwin-framework-tool to exercise it.  We should not be adding things just so they can be used from darwin-framework-tool.

bzbarsky-apple commented:
Also, note the memory leaks this is introducing (hence some of the CI failures).

seokhee-lee commented:
>

seokhee-lee commented:
I removed all discovery and ethernet pairing changes because they are useful for test but not a mandatory feature for Matter certification. I updated the commit and description to focus on "read event" function needed for Matter Certification of controller app.

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -241,3 +241,68 @@ class SubscribeEvent : public ModelCommand {
+    {
+        AddArgument("cluster-id", 0, UINT32_MAX, &mClusterId);
+        AddArgument("event-id", 0, UINT32_MAX, &mEventId);
+        AddArgument("fabric-filtered", 0, 1, &mFabricFiltered);
```
This part looks like it's duplicating chip-tool, but event reads ignore this fabric-filtered boolean, no?  So we should not have it here, for any of these functions.

seokhee-lee replied:
In my understanding, the fabric-filter option is needed for all matter commands including read event. We tested that the option is working in read event scenario. If it is misunderstanding of the option, please let me know. I will remove the option in the functions. Our test sequence is as follows:

-Pre-condition
     1. Paring is completed between lock-app and darwin-framework-tool.
     2. Generate door-lock-alarm event with "echo '{"Cmd": "SendDoorLockAlarm", "Params": { "EndpointId": 1, "AlarmCode": 0 } }' > /tmp/chip_lock_app_fifo-<PID>   (PID of lock app)"

-Test sequence
    1. Read door-lock-alarm event with fabric-filter option 0 or 1: "doorlock read-event-by-id 0 123456 1 --fabric-filtered 0 or 1"
   2. Check lock-app log whether isFabricFiltered value in ReadRequestMessage is set as expected: 0 -> false and 1 -> true.

bzbarsky-apple replied:
The "fabric filtered" option only affects the behavior of reading:

1. Attributes
2. That are of type list.
3. Where the entries in the list are structs
4. Where the structs are fabric-scoped.

It has no impact on reading events at all.  That is, you can send the boolean in the request message, but it will not change what data you get back as a result...

seokhee-lee replied:
Maybe it seems my misunderstanding. I have thought that the option affects the other fabric because chip-tool read-event also has that option. I will double check the meaning of the option and actual result in multi-fabric environments. I will update soon after that.

seokhee-lee replied:
@bzbarsky-apple I removed fabric-filtered option in read-event-by-id of darwin-framework-tool.

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -241,3 +241,68 @@ class SubscribeEvent : public ModelCommand {
+                                  }
+                                  if (values) {
+                                      for (id item in values) {
+                                          NSLog(@"Response Item: %@", [item description]);
```
This doesn't seem "logging-compatible" with the "read-event" bits in chip-tool.  Is it expected to be?  Will certification expect it to be?

seokhee-lee replied:
I am just following the existing darwin-framework-tool logging style for it. Because we will use our iOS controller app log in certification, the darwin-framework-tool log is not important for certification itself. I just implement it as example to use the new matter framework API "readEventsWithEndpointID". If some modification is needed, please let me know.

bzbarsky-apple replied:
Please check with @krypton36 what he wants for darwin-framework-tool here?

seokhee-lee replied:
@krypton36 could you give some advice of this code change? I want to show how to use the new API readEventsWithEndpointID to the other app developers and to verify my implementation of it by using darwin-framework-tool.

bzbarsky-apple commented:
@seokhee-lee You should probably also fix the restyle issues, so this will request reviews more widely...

joonhaengHeo commented:
@seokhee-lee If you check restyle, you can use below script.
./scripts/helpers/restyle-diff.sh

joonhaengHeo commented:
@seokhee-lee
If you look at your modifications, the zap file modifications are included.
In this case, the item "ZAP/ZAP templates generation (pull_request)" will not fail until the zap file is recreated.

joonhaengHeo commented:
> > @seokhee-lee If you look at your modifications, the zap file modifications are included. In this case, the item "ZAP/ZAP templates generation (pull_request)" will not fail until the zap file is recreated.
> 
> @joonhaengHeo I will commit Commands.h generated by the changed zapt file after fixing restyle error of it. Could you guide how to generate file with zapt without restyle error? I fixed the restyle errors except the auto-generated file and the restyling tool is not working for the file for now because it seems to have too many changes.

@seokhee-lee Don't you check the restyle for files created using zap?

seokhee-lee commented:
> 

@joonhaengHeo I uploaded the generated file after restyling. There is no restyling error but zap templates generation error does not disappear. Could you let me know what I should do more?

seokhee-lee commented:
> > 
> 
> @joonhaengHeo I uploaded the generated file after restyling. There is no restyling error but zap templates generation error does not disappear. Could you let me know what I should do more?

I fixed the zap-related issue. I did not know that the output location of zap-cli affects the generated file's formate.

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -206,27 +221,51 @@ class SubscribeEvent : public ModelCommand {
-            eventReportHandler:^(NSArray * value) {
-                for (id item in value) {
-                    NSLog(@"Response Item: %@", [item description]);
+        if (mClusterId != chip::kInvalidAttributeId) {
```
Why we are comparing a cluster id to an attribute id?

seokhee-lee replied:
It changes for mClusterId to be compared to kInvalidClusterId.

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -237,7 +276,88 @@ class SubscribeEvent : public ModelCommand {
+    bool DeferInteractiveCleanup() override { return mSubscriptionEstablished; }
+
+private:
+    chip::ClusterId mClusterId = chip::kInvalidAttributeId;
```
Why are we setting a cluster id to an attribute id?

seokhee-lee replied:
It changes for mClusterId to be set to kInvalidClusterId.

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -237,7 +276,88 @@ class SubscribeEvent : public ModelCommand {
+    ReadEvent(const char * _Nonnull eventName)
+        : ModelCommand("read-event")
+    {
+        AddArgument("event-name", eventName);
```
Where does this set mClusterId or mEventId?  I'm not clear on how this constructor ends up working, at least to set the members we use below. I also don't see it actually being used in the codegen.

seokhee-lee replied:
All arguments including mClusterId and mEventId are set in InitArgument method of Command class which is one of the parent classes of ReadEvent. If user input of darwin-framework-tool is including read-event-by-id, ReadEvent instance generates and all member variables are set by initArgument method based on the user input before calling SendCommand of ReadEvent.

seokhee-lee replied:
"read-event" command is deleted and it will be implemented later if app developer has some needs (now only read-event-by-id is being used in our side).

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.h:
```diff
@@ -315,6 +315,50 @@ typedef NS_ENUM(uint8_t, MTRTransportType) {
+- (void)readEventsWithEndpointID:(NSNumber * _Nullable)endpointID
+                       clusterID:(NSNumber * _Nullable)clusterID
+                         eventID:(NSNumber * _Nullable)eventID
+                        eventMin:(NSNumber * _Nullable)eventMin
```
I'd argue eventMin should be in MTRReadParams, not a separate argument here.  And it should document what the behavior is if it's left nil.

seokhee-lee replied:
eventMin is into MTRReadParams and documented based on Matter 1.0 Specification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.h:
```diff
@@ -315,6 +315,50 @@ typedef NS_ENUM(uint8_t, MTRTransportType) {
+                              clusterID:(NSNumber * _Nullable)clusterID
+                                eventID:(NSNumber * _Nullable)eventID
+                               eventMin:(NSNumber * _Nullable)eventMin
+                               isUrgent:(NSNumber * _Nullable)isUrgent
```
isUrgent should (1) be a BOOL, I would think, and (2) be on MTRSubscribeParams.  Again, the behavior for it (e.g. the default value) needs to be documented.

seokhee-lee replied:
isUrgentEvent is into MTRSubscribeParams with BOOL type and documented based on Matter 1.0 Specification.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1470,6 +1471,327 @@ + (id)CHIPEncodeAndDecodeNSObject:(id)object
     return decodedData.GetDecodedObject();
 }
 
+template <typename DecodableEventType> class BufferedReadEventCallback final : public app::ReadClient::Callback {
```
Just to repeat again: this is duplicating a _lot_ of the attribute code.  Why are we not sharing that code, after a bit of refactoring, instead of copying it?

seokhee-lee replied:
Added BufferedReadClientCallback instead of BufferedReadAttributeCallback and BufferedReadEventCallback to reduce duplication.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1470,6 +1471,327 @@ + (id)CHIPEncodeAndDecodeNSObject:(id)object
+            auto resultArray = [[NSMutableArray alloc] init];
+            auto resultSuccess = [[NSMutableArray alloc] init];
+            auto resultFailure = [[NSMutableArray alloc] init];
+            auto onSuccessCb = [resultArray, resultSuccess](
```
And this logic seems to duplicate a lot of the readAttribute logic...

seokhee-lee replied:
To reduce the duplication, we integrate two buffered read callbacks for attribute and event into one. We also agree that it is importance to enhance code quality to minimize duplication and to enhance readability/reusability. In order to minimize duplication, I think that read and subscribe can be integrated into one function like linux case and there seems to exist some other approaches which we can try. We will submit PR for them after this PR is merged.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1470,6 +1471,327 @@ + (id)CHIPEncodeAndDecodeNSObject:(id)object
+            [params toReadPrepareParams:readParams];
+            readParams.mpEventPathParamsList = &eventPath;
+            readParams.mEventPathParamsListSize = 1;
+            readParams.mEventNumber.SetValue(static_cast<chip::EventNumber>([eventMin unsignedLongLongValue]));
```
Does this do the right thing when `eventMin` is nil?  I guess it does because we will get 0, but more correct would be to not set mEventNumber at all.

seokhee-lee replied:
We changed for Read Request to be sent without EventFilterIB if eventMin is nil.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1470,6 +1471,327 @@ + (id)CHIPEncodeAndDecodeNSObject:(id)object
+        getSessionForNode:self.nodeID
+               completion:^(ExchangeManager * _Nullable exchangeManager, const Optional<SessionHandle> & session,
+                   NSError * _Nullable error) {
+                   if (error != nil) {
```
Again, this is duplicating a lot of the attribute code... Not sure whether there's a good way to share things here, though.

seokhee-lee replied:
To reduce the duplication, we integrate two buffered read callbacks for attribute and event into one. In order to minimize duplication, I think that read and subscribe can be integrated into one function like linux case. We will submit PR for it after this PR is merged because It seems that lots of code changes are needed for the integration.

bzbarsky-apple commented:
@ready2die4u I am sorry I did not get to this review yet...  I will make sure I do that on Tuesday (Monday is a holiday in the US).

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -199,34 +214,59 @@ class SubscribeEvent : public ModelCommand {
-                for (id item in value) {
-                    NSLog(@"Response Item: %@", [item description]);
+        if (mClusterId != chip::kInvalidClusterId) {
+            [device subscribeToEventsWithEndpointID:[NSNumber numberWithUnsignedShort:endpointId]
+                clusterID:[NSNumber numberWithUnsignedInteger:mClusterId]
+                eventID:[NSNumber numberWithUnsignedInteger:mEventId]
```
If this is meant to be API-compatible with chip-tool, this should probably treat the invalid id values as wildcards...  Followup for doing that, please.

seokhee-lee replied:
@bzbarsky-apple This is current status. If more changes are needed, please let us know:
1. Arguments of 0xFFFF is working as wildcards like chip-tool
2. To support more wildcard logics same to chip-tool, "any" cluster is added

bzbarsky-apple replied:
> Arguments of 0xFFFF is working as wildcards like chip-tool

Sort of by accident, because the Darwin framework does not reject them as invalid values.  Which it probably should, because those are not valid ids.  The way wildcards are represented in the Darwin API is with `nil`, not magic invalid values....

bzbarsky-apple commented on examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:
```diff
@@ -237,7 +277,81 @@ class SubscribeEvent : public ModelCommand {
+            params.eventMin = [NSNumber numberWithUnsignedLongLong:mEventNumber.Value()];
+        }
+
+        [device readEventsWithEndpointID:[NSNumber numberWithUnsignedShort:endpointId]
+                               clusterID:[NSNumber numberWithUnsignedInteger:mClusterId]
+                                 eventID:[NSNumber numberWithUnsignedInteger:mEventId]
```
Again, the wildcard handling here is off.  Same followup as above.

seokhee-lee replied:
@bzbarsky-apple This is current status. If more changes are needed, please let us know:
1. Arguments of 0xFFFF is working as wildcards like chip-tool
2. To support more wildcard logics same to chip-tool, "any" cluster is added

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.h:
```diff
@@ -315,6 +315,47 @@ typedef NS_ENUM(uint8_t, MTRTransportType) {
+ * (e.g. nil eventID means "read all the events from the endpoint(s) and
+ * cluster(s) that match endpointID/clusterID").
+ *
+ * If all of endpointID, clusterID, eventID are non-nil, a single
+ * event will be read.
```
```suggestion
 * If all of endpointID, clusterID, eventID are non-nil, all the matching instances of a single
 * event will be read.
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1702,6 +1970,13 @@ - (instancetype)initWithPath:(const ConcreteEventPath &)path
+- (NSString *)description
+{
+    return
+        [NSString stringWithFormat:@"<MTREventPath> endpoint %u cluster %u event %u", (uint16_t) self.endpoint.unsignedShortValue,
+                  (uint32_t) self.cluster.unsignedLongValue, (uint32_t) _event.unsignedLongValue];
```
Why are the casts here needed?  They should not be needed I would think.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -94,11 +94,21 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, assign, getter=shouldFilterByFabric) BOOL filterByFabric MTR_NEWLY_AVAILABLE;
 
+/**
+ * Sets the EventMin filed in the EventFilterIB for the read/subscribe for the interaction.
```
```suggestion
 * Sets a filter for which events will be reported in the read/subscribe interaction.
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -94,11 +94,21 @@ NS_ASSUME_NONNULL_BEGIN
+ *
+ * If nil (the default value), all of the queued events will be reported from lowest to highest event number.
+ *
+ * If not nil, the queued events will be reported except for ones that the event number is less than the EventMin field in the
+ * EventFilterIB.
```
```suggestion
 * If not nil, queued events with an event number smaller than eventMin will not be reported.
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -144,6 +154,15 @@ NS_ASSUME_NONNULL_BEGIN
+ *
+ * If NO, the events will be reported at the maximum interval.
+ */
+@property (nonatomic, assign, getter=isUrgentEvent) BOOL isUrgentEvent;
```
```suggestion
@property (nonatomic, assign) BOOL reportEventsUrgently MTR_NEWLY_AVAILABLE;
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -144,6 +154,15 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, copy) NSNumber * maxInterval;
 
+/**
+ * Sets IsUrgent flag of EventPathIB in the Subscribe Request. The default value is NO.
```
```suggestion
 * Controls whether events will be reported urgently. The default value is YES.
```

bzbarsky-apple replied:
Default needs to be YES to match current behavior of `subscribeWithQueue`.

Also, we should be using this value in `subscribeWithQeuue`, now that we have it, and remove the hardcoded `true` there.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -144,6 +154,15 @@ NS_ASSUME_NONNULL_BEGIN
+/**
+ * Sets IsUrgent flag of EventPathIB in the Subscribe Request. The default value is NO.
+ *
+ * If YES, the events will be reported as soon as queued.
```
```suggestion
 * If YES, the events will be reported as soon as the minInterval does not prevent it.
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.mm:
```diff
@@ -88,6 +92,7 @@ @implementation MTRSubscribeParams
 - (instancetype)initWithMinInterval:(NSNumber *)minInterval maxInterval:(NSNumber *)maxInterval
 {
     if (self = [super init]) {
+        _isUrgentEvent = NO;
```
```suggestion
        _isUrgentEvent = YES;
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -94,11 +94,21 @@ NS_ASSUME_NONNULL_BEGIN
+ * If not nil, the queued events will be reported except for ones that the event number is less than the EventMin field in the
+ * EventFilterIB.
+ */
+@property (nonatomic, copy, nullable) NSNumber * eventMin MTR_NEWLY_AVAILABLE;
```
```suggestion
@property (nonatomic, copy, nullable) NSNumber * minimumEventNumber MTR_NEWLY_AVAILABLE;
```

bzbarsky-apple reviewed (approved):
I'm sorry this took so long....  Thank you for being so patient!

bzbarsky-apple commented:
@ready2die4u @joonhaengHeo I feel bad asking you to hurry after it took me so long to get to the review, but if we can iron out the final few bits and get this merged in the next few days that would make it much more likely that it appears in a system Matter.framework in the near future.

One thing that I suspect we do need here that I still don't see, by the way, is tests for the new code...

seokhee-lee commented:
> @ready2die4u @joonhaengHeo I feel bad asking you to hurry after it took me so long to get to the review, but if we can iron out the final few bits and get this merged in the next few days that would make it much more likely that it appears in a system Matter.framework in the near future.
> 
> One thing that I suspect we do need here that I still don't see, by the way, is tests for the new code...

@bzbarsky-apple , I thank you for the reviews and sorry for the late reply because there was big holiday (New Year's Day in lunar calendar) in Korea. We will update the patches by tomorrow according to your reviews. We are testing the codes with darwin-framework-tool described in the commit message and our controller iOS app. In addition, all commands of darwin-framework-tool are being tested to guarantee that our codes do not cause any side effects to the existing codes. If we need more works to verify our code changes, please let me know.

bzbarsky-apple commented:
@seokhee-lee I totally understand about the holiday!

In addition to the manual tests, these APIs need automated tests (see existing tests in MTRDeviceTests.m for examples).   That can be a follow-up PR as long as it happens.

ready2die4u commented:
I'm sorry for delay. I have changed it as shown below:
- Modified the documents about API and renamed the property
- Modified the allocator of SubscribeEvent class for 'Any' cluster to  add subscribe-event-by-id to commands.

I will prepare a follow-up PR for automated tests.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.mm:
```diff
@@ -111,6 +118,9 @@ - (void)toReadPrepareParams:(chip::app::ReadPrepareParams &)readPrepareParams
     readPrepareParams.mMinIntervalFloorSeconds = self.minInterval.unsignedShortValue;
     readPrepareParams.mMaxIntervalCeilingSeconds = self.maxInterval.unsignedShortValue;
     readPrepareParams.mKeepSubscriptions = !self.replaceExistingSubscriptions;
+    if (self.minimumEventNumber) {
+        readPrepareParams.mEventNumber.SetValue(static_cast<chip::EventNumber>([self.minimumEventNumber unsignedLongLongValue]));
+    }
```
This is handled by the super call above, no?  Sorry, I should have caught this last time.

seokhee-lee replied:
@bzbarsky-apple we removed the code because it is duplicated as you mentioned.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1707,6 +1975,13 @@ - (instancetype)initWithPath:(const ConcreteEventPath &)path
+- (NSString *)description
+{
+    return
+        [NSString stringWithFormat:@"<MTREventPath> endpoint %u cluster %u event %u", (uint16_t) self.endpoint.unsignedShortValue,
+                  (uint32_t) self.cluster.unsignedLongValue, (uint32_t) _event.unsignedLongValue];
```
My question about casts here is still unresolved.

seokhee-lee replied:
@bzbarsky-apple if the casts are not used here, we found that build error during compiling matter framework occurs.

bzbarsky-apple replied:
Ok, sounds good.  I will sort this out later.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRCluster.h:
```diff
@@ -94,11 +94,20 @@ NS_ASSUME_NONNULL_BEGIN
+ *
+ * If nil (the default value), all of the queued events will be reported from lowest to highest event number.
+ *
+ * If not nil, queued events with an event number smaller than eventMin will not be reported.
```
```suggestion
 * If not nil, queued events with an event number smaller than minimumEventNumber will not be reported.
```

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -1475,6 +1501,248 @@ + (id)CHIPEncodeAndDecodeNSObject:(id)object
+                   if (eventID) {
+                       container.eventPathParams->mEventId = static_cast<chip::EventId>([eventID unsignedLongValue]);
+                   }
+                   if (params.reportEventsUrgently) {
+                       container.eventPathParams->mIsUrgentEvent = params.reportEventsUrgently;
+                   }
```
```suggestion
                   container.eventPathParams->mIsUrgentEvent = params.reportEventsUrgently;
```

bzbarsky-apple reviewed (commented):
The following issues from the previous review are not addressed:

1. https://github.com/project-chip/connectedhomeip/pull/24057#discussion_r1083239363
2. The part about `subscribeWithQueue` in https://github.com/project-chip/connectedhomeip/pull/24057#discussion_r1083239880
3. Filing followups for the wildcard handling as described in https://github.com/project-chip/connectedhomeip/pull/24057#discussion_r1083237517 and https://github.com/project-chip/connectedhomeip/pull/24057#discussion_r1083237606.  This does not block landing, but please make sure the issues are filed.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -338,7 +338,6 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
                                       auto attributePath = std::make_unique<AttributePathParams>();
                                       auto eventPath = std::make_unique<EventPathParams>();
                                       // We want to get event reports at the minInterval, not the maxInterval.
-                                      eventPath->mIsUrgentEvent = true;
```
This needs to get set from `params`, no?  Something like:
````
    eventPath->mIsUrgentEvent = params.reportEventsUrgently;
```

otherwise this is completely losing the urgent-event bits in this function...

bzbarsky-apple replied:
I'm just going to fix this up, in the interests of getting this merged.

bzbarsky-apple commented:
Fast-tracking platform-specific change with platform owner review.

bzbarsky-apple commented:
@joonhaengHeo @ready2die4u Thank you again for your patience, and sorry this took so long!

I filed https://github.com/project-chip/connectedhomeip/issues/24666 to track adding automated tests for these APIs.

seokhee-lee commented:
> @joonhaengHeo @ready2die4u Thank you again for your patience, and sorry this took so long!
> 
> I filed #24666 to track adding automated tests for these APIs.

@bzbarsky-apple I thank you for your effort to review and merge the changes. I will commit the automated test patches soon.
